### PR TITLE
Improve mac parallel performance

### PR DESCRIPTION
**Pull request by dhiltgen**
Carries #3900 

Testing underway...

### CODE REVIEW

File: server/sched.go [hunk 1]
```diff
@@ -81,6 +90,10 @@ func (s *Scheduler) GetRunner(c context.Context, model *Model, opts api.Options,
 		successCh:       make(chan *runnerRef),
 		errCh:           make(chan error, 1),
 	}
+	if numParallel > 1 {
```
Comment (bsdnet) [inline server/sched.go:93]: Do we need "if" in this case?

File: llm/patches/04-metal.diff [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
++++ b/ggml-metal.m
+@@ -1396,27 +1396,23 @@ static enum ggml_status ggml_metal_graph_compute(
+                         // to the matrix-vector kernel
+                         int ne11_mm_min = 1;
+ 
+-#if 0
+                         // the numbers below are measured on M2 Ultra for 7B and 13B models
+                         // these numbers do not translate to other devices or model sizes
+                         // TODO: need to find a better approach
+-                        if ([ctx->device.name isEqualToString:@"Apple M2 Ultra"]) {
```
Comment (BruceMacD) [inline llm/patches/04-metal.diff:13]: Are we doing test coverage on an M1 for this? Curious why this was originally being bypassed on older hardware.

File: server/sched.go [hunk 3]
```diff
@@ -57,6 +58,14 @@ func InitScheduler(ctx context.Context) *Scheduler {
 			loadedMax = m
 		}
 	}
+	if onp := os.Getenv("OLLAMA_NUM_PARALLEL"); onp != "" {
+		p, err := strconv.Atoi(onp)
+		if err != nil || p <= 0 {
+			slog.Error("invalid parallel setting, must be greater than zero", "OLLAMA_NUM_PARALLEL", onp, "error", err)
```
Comment (BruceMacD) [inline server/sched.go:64]: Should we fail here since the server wont act as they expect?

File: server/sched.go [hunk 4]
```diff
@@ -81,6 +90,8 @@ func (s *Scheduler) GetRunner(c context.Context, model *Model, opts api.Options,
 		successCh:       make(chan *runnerRef),
 		errCh:           make(chan error, 1),
 	}
+	// context split across parallel threads
+	opts.NumCtx = opts.NumCtx * numParallel
```
Comment (BruceMacD) [inline server/sched.go:94]: I might not be interpreting this incorrectly, but this looks like it assumes both models running in parallel have the same context size, which isn't going to be true a lot of the time.

File: server/sched.go (hunk 3 above)
Comment (jmorganca) [reply server/sched.go:64]: I think it's ok to just log an error (vs a total crash)

File: server/sched.go (hunk 4 above)
Comment (jmorganca) [reply server/sched.go:94]: This is one model, many parallel requests. llama.cpp will split the kv cache between all parallel requests so we need to make the context size n times bigger

Comment (BruceMacD) [reply server/sched.go:94]: I see, thanks for the clarification

**Review by jmorganca — APPROVED** [review]
Overall testing seems good with num_batch=512. Less thorough testing of num_batch < 512 still shows good performance

File: llm/patches/04-metal.diff (hunk 2 above)
Comment (BruceMacD) [reply llm/patches/04-metal.diff:13]: Looks like Jeff got this

**Review by BruceMacD — APPROVED** [review]

**Pull request merged**